### PR TITLE
Enable DamageInd Angle in Client

### LINKED ISSUE #2141: Damage Indicators
Server sending unused ClientID, Angle and Self variables
https://github.com/teeworlds/teeworlds/blob/d18b5c3fec3b4c028b2c10af850206fdec6a1621/src/game/server/gamecontext.cpp#L96-L110

Happened after this(?) https://github.com/teeworlds/teeworlds/commit/029b0accbd7eebc0e3f8d205f62ff6ccf9da77b6

### LINKED ISSUE #2787: damage indicator angle value not used
would be cool to allow server to send an angle again without the client just ignoring it

### PR DESCRIPTION
**Pull request by TsFreddie**
closes #2787

TLDR; the angle is enabled but it looks different (duh).

Not sure if the laser's damage angle is intended but it is being sent like this, I didn't modify any server code.

A pretty big change is that I removed the code to layout the stars in a circle, which removes the effect where the stars goes around the tee when you are quickly firing pistols towards a **stationary tee**, that effect is not very visible when tees are moving anyway. Furthermore, it was on a global timer so when the fight is really hectic the stars may seemingly come from all over the place.

Seems like the intention of that piece of code was to stop shotgun damage stars from grouping together. I used the ClientID (thanks Dune) to process all damage events within a snapshot before creating stars, so the shotgun damage is laid out pretty nicely.

In cases where the shotgun is fired towards a tee from far distance. The bullets might not hit the target within a tick hence creating two or three groups of stars, although looks weird, but the behaviour is actually in line with hit sounds where the sound would play multiple times and give you one set of stars per sound. (Plus, the fact that the stars from shotguns are usually in 1 + 2 formation and usually don't collides makes this behaviour looks fine)

In rare cases where multiple attackers hitting a single target within one tick, the stars will layout in one single group with the angle averaged out, which is harder to solve with no attacker info. But it means the fight is probably chaotic enough for this to happen regularly where the grouped damage might be better to look at.

Preview (ignore the wacky skin marking please):
image
image
image
image
image

_Sidenote: modders can send stars using a ClientID of -1 to simulate 0.6's star behaviour (i.e. stars wont group at all for client -1)._

### CODE REVIEW

File: src/game/client/components/effects.h [hunk 1]
```diff
@@ -6,16 +6,25 @@
 
 class CEffects : public CComponent
 {
+	struct CDamageIndInfo
+	{
+		vec2 Pos;
```
Comment (fokkonaut) [inline src/game/client/components/effects.h:11]: m_ prefixes

**Comment by heinrich5991** [comment]
> A pretty big change is that I removed the code to layout the stars in a circle, which removes the effect where the stars goes around the tee when you are quickly firing pistols towards a **stationary tee**, that effect is not very visible when tees are moving anyway. Furthermore, it was on a global timer so when the fight is really hectic the stars may seemingly come from all over the place.

That's a bit sad, I liked that effect. You didn't tell us why you removed it though. Code complexity?

**Comment by TsFreddie** [comment]
> > A pretty big change is that I removed the code to layout the stars in a circle, which removes the effect where the stars goes around the tee when you are quickly firing pistols towards a **stationary tee**, that effect is not very visible when tees are moving anyway. Furthermore, it was on a global timer so when the fight is really hectic the stars may seemingly come from all over the place.
> 
> That's a bit sad, I liked that effect. You didn't tell us why you removed it though. Code complexity?

mostly because it does not make sense without a fixed starting angle now that it used angles sent by server. Plus the fact you can't really see it if you aren't just testing it with stationary tees.

For example, in the old method, you have a fixed starting angle of about 270 degrees. Each damage will generate a new star counterclockwise no matter which direction you've been hit from. Now, with hit angle considered, if we get 1 damage hit from the 0 degree angle multiple times. We would want stars in those directions because it now convey direction info's. If we stack them the same way (with some angle info stored and checked since a hit from 180 degree shouldn't group with the -10~10 degrees ones), the direction information it suppose to convey breaks.

Now. Solution one, we CAN stack them with some animation that moves the original stars clockwise to keep the entire star group facing the general direction (like a shotgun damage). But when the tees are moving, you'll get stars that seemingly moves on its own because the grouping visual is gone given the stars that supposed to group have different positions.

Solution two is to stack stars in the same position as the first star instead of using position sent by server, but this time breaks position info.

Solution three is to stop using position info and attach the stars to the player. Well, see it for yourself: https://cdn.discordapp.com/attachments/510879693982007316/770844961864024114/2020-10-28_11-00-56.mp4
The stars doesn't stack yet but this is already pretty hard to look at.

Solution four is to detect whether the tee is moving, if it is we don't stack, if it isn't we stack them with animation. We then have two different behaviors of the stars.

Solution five is to ditch the current burst animation and design a new one. Well, probably still not the same as the old stacking one.

Solution six is we give up on the issues. Bring back the non directional stars. (Which looks the same as 0.6 anyway). And maybe fix the global timer thing and enable the direction for mods using a clientid check. Which may not be a bad idea but it is still two different behaviors in one design.

Finally we can just keep the old one, which is why this post and the op one is so long to provide enough context for everyone to consider.

And if someone already had some animation in mind, now it is probably a good time to bring up.

**Comment by TsFreddie** [comment]
Well a simpler reason may be it really only work in one condition and makes something like fng timer and ddr pistol really hard to look at.

I do love there are so many things to consider when it comes to a simple effect like this tho.

**Comment by Dune-jr** [comment]
Doesn't this mean the stars may stack? The whole point of the circle effect was to avoid this stacking effect (as it doesn't look great and may be difficult to see/count)

> closes #2787 and closes #2141

By the way, this PR does not really close https://github.com/teeworlds/teeworlds/issues/2141. There is material to improve much more than the way the angle is being set, especially make use of `HealthAmount`/`ArmorAmount` to subtly differentiate pure armor damage from the rest with some refined animations.

**Comment by TsFreddie** [comment]
> Doesn't this mean the stars may stack? The whole point of the circle effect was to avoid this stacking effect (as it doesn't look great and may be difficult to see/count)
> 
> > closes #2787 and closes #2141
> 
> By the way, this PR does not really close https://github.com/teeworlds/teeworlds/issues/2141. There is material to improve much more than the way the angle is being set, especially make use of `HealthAmount`/`ArmorAmount` to subtly differentiate pure armor damage from the rest with some refined animations.

ok, stack is a bad word choice on my part. By stacking I'm not saying they overlap, more like they orderly stacking like marbles in a tube.

**Comment by TsFreddie** [comment]
Let's go again.

> I guess I'll try some other animation to avoid overlapping stars on stationary tees.

Didn't bring the circle thing back, for reasons mentioned above. But just for fun here's a video of the circle thing with server angle enabled, and it shows why the direction information would be lost by the end of attacking:
https://cdn.discordapp.com/attachments/510879693982007316/771245867470290944/2020-10-29_13-31-58.mp4

Instead, I made a fixed pattern for quick consecutive damages.
https://cdn.discordapp.com/attachments/510879693982007316/771244258002075658/2020-10-29_13-28-51.mp4

It runs on a timer for individual client and resets quickly so you can only trigger the pattern with pistol shots.

How do you like this as a compromise @heinrich5991 😄
